source/application/AMPIC887DriverServer/AMPIC887AxisMap.h

```diff
@@ -0,0 +1,297 @@
+#include "util/AMCArrayHandler.h"
+/*!
+  * A class which maps Axes of a PI C887.11 controller to a type T.
+  */
+template <typename T>
+class AMPIC887AxisMap : public QMap<AMGCS2::Axis, T>
+{
+public:
+	/*!
+	  * Creates an instance of an axis map, which is empty.
+	  */
+	AMPIC887AxisMap() {}
+
+	/*!
+	  * Returns a collection of the axes which this map contains.
+	  */
+	AMPIC887AxisCollection axes() const
+	{
+		AMPIC887AxisCollection returnCollection(AMPIC887AxisCollection::EmptyCollection);
+		QList<AMGCS2::Axis> axisEntries = this->keys();
+
+		foreach(AMGCS2::Axis currentAxisEntry, axisEntries) {
+			returnCollection.append(currentAxisEntry);
+		}
+
+		return returnCollection;
+	}
+
+	/*!
+	  * Outputs the map values as a human readable string, if it can parse the
+	  * type into a string, otherwise returns the empty string.
+	  */
+	QString toString(const QString& = "Data") const
+	{
+		// Implemented for known types in partial specializations below.
+		return QString();
+	}
+
+	/*!
+	  * Whether this map contains an entry for the UnknownAxis.
+	  */
+	bool containsUnknownAxis() const
+	{
+		return this->contains(AMGCS2::UnknownAxis);
+	}
+
+	/*!
+	  * A map of the provided axes to their values stored within this map. If
+	  * a key contained within the provided collection is not contained within
+	  * this map it will not be included in the returned map.
+	  * \param keys ~ A collection of axes whose values within this map will
+	  * be returned.
+	  */
+	AMPIC887AxisMap<T> values(const AMPIC887AxisCollection& keys) const
+	{
+		AMPIC887AxisMap<T> returnValues;
+		foreach(AMGCS2::Axis currentAxis, keys) {
+			if(this->contains(currentAxis)) {
+				returnValues.insert(currentAxis, this->value(currentAxis));
+			}
+		}
+
+		return returnValues;
+	}
+
+};
+
+// Partial Specialization on type T to provide functionality of toString() method
+////////////////////////////////////////////////////////////////////////////////
+
+template <>
+class AMPIC887AxisMap<bool> : public QMap<AMGCS2::Axis, bool>
+{
+public:
+	/*!
+	  * Creates an instance of an axis map, which is empty.
+	  */
+	AMPIC887AxisMap() {}
+
+	/*!
+	  * Returns a collection of the axes which this map contains.
+	  */
+	AMPIC887AxisCollection axes() const
+	{
+		AMPIC887AxisCollection returnCollection(AMPIC887AxisCollection::EmptyCollection);
+		QList<AMGCS2::Axis> axisEntries = this->keys();
+
+		foreach(AMGCS2::Axis currentAxisEntry, axisEntries) {
+			returnCollection.append(currentAxisEntry);
+		}
+
+		return returnCollection;
+	}
+
+	/*!
+	  * Outputs the map values as a human readable string, if it can parse the
+	  * type into a string, otherwise returns the empty string.
+	  */
+	QString toString(const QString& header = "Data") const
+	{
+		QString returnString = QString("Axis\t\t%1\n").arg(header);
+
+		QList<AMGCS2::Axis> axes = this->keys();
+		foreach(AMGCS2::Axis currentAxis, axes) {
+			QString axisValue;
+			if(this->value(currentAxis)) {
+				axisValue = "Yes";
```

Comment:
This whole loop could probably be reduced to a single line of code.  You don't have to make the modification, but I thought I'd bring it up.